feat(erc721): return Vec<u8> on IErc721Wrapper function calls

sumitvekariya opened the pull request:
## Description

This PR aligns the ERC721Wrapper error handling behavior with Solidity by making the `IErc721Wrapper` trait return `Vec<u8>` on all function calls instead of structured Error types. This change ensures that underlying token call errors are propagated directly as raw bytes, maintaining consistency with the Solidity reference implementation.

## Problem

Previously, the `IErc721Wrapper` trait methods (`deposit_for`, `withdraw_to`, and `on_erc721_received`) returned generic `Result<T, Error>` types. This approach differed from the Solidity behavior where underlying token errors are propagated directly to the caller as raw revert data.

## Solution

- **Updated `IErc721Wrapper` trait interface**: Changed return types from `Result<bool, Error>` to `Result<bool, Vec<u8>>` for `deposit_for` and `withdraw_to` methods
- **Modified error propagation**: Used `MethodError::encode()` to convert underlying call errors to raw bytes (`Vec<u8>`)
- **Updated `on_erc721_received`**: Changed return type from `Result<B32, Error>` to `Result<B32, Vec<u8>>`
- **Enhanced documentation**: Updated method documentation to clarify that raw revert data is returned to align with Solidity behavior

## Changes Made

### Core Implementation (`contracts/src/token/erc721/extensions/wrapper.rs`)

1. **Updated trait interface**:
   - `deposit_for`: `Result<bool, Error>` → `Result<bool, Vec<u8>>`
   - `withdraw_to`: `Result<bool, Error>` → `Result<bool, Vec<u8>>`
   - `on_erc721_received`: Generic error type → `Result<B32, Vec<u8>>`

2. **Modified implementation methods**:
   - Replaced `|e| e.encode()` closures with direct `MethodError::encode` calls (fixing clippy warnings)
   - Updated error handling to propagate underlying token call errors as raw bytes
   - Maintained existing logic while changing return type handling

3. **Updated comprehensive test suite**:
   - Changed test assertions from expecting `Error` types to expecting `Vec<u8>`
   - Updated success case expectations to use proper byte representations
   - All 30 wrapper-related tests pass successfully

### Example Implementation (`examples/erc721-wrapper/src/lib.rs`)

- Updated example implementation to match new trait interface
- Fixed method signatures to return `Result<bool, Vec<u8>>`
- Removed unnecessary wrapper error imports

### Documentation (`docs/modules/ROOT/pages/erc721-wrapper.adoc`)

- Updated documentation examples to reflect new interface
- Added clarifications about raw error data propagation
- Maintained comprehensive usage examples

## Testing

- ✅ All existing functionality preserved
- ✅ 30 ERC721Wrapper tests pass
- ✅ Example implementation compiles and works correctly  
- ✅ Documentation examples updated and validated
- ✅ Clippy warnings resolved (redundant closures fixed)

## Behavioral Changes

- **Error Handling**: Underlying token call errors are now propagated as raw bytes instead of structured error types
- **Return Types**: Method signatures changed from generic error types to `Vec<u8>` 
- **Solidity Alignment**: Error propagation now matches Solidity behavior exactly

## Breaking Changes

⚠️ **This is a breaking change** for users of the `IErc721Wrapper` trait:
- Method return types have changed
- Error handling patterns need to be updated
- Test assertions expecting structured errors need to be updated to expect `Vec<u8>`

## Migration Guide

**Before:**
```rust
match wrapper.deposit_for(account, token_ids) {
    Err(Error::InsufficientApproval(_)) => {
        // Handle structured error
    }
}
```

**After:**
```rust
match wrapper.deposit_for(account, token_ids) {
    Err(error_bytes) => {
        // Handle raw error bytes - can decode if needed
        let expected_error: Vec<u8> = Error::InsufficientApproval(/* ... */).into();
        assert_eq!(error_bytes, expected_error);
    }
}
```

## Closes

Fixes #801

---

This PR ensures that the Rust implementation of ERC721Wrapper behaves consistently with its Solidity counterpart, providing better interoperability and maintaining expected error propagation patterns.

0xNeshi commented:
Hi @sumitvekariya , thanks for the contribution! 
As we're in the middle of an audit, we need to focus on tackling related issues. We will make sure to review your PR as soon as possible.
Thanks once again!

qalisander commented:
We need update README.md

0xNeshi reviewed (approved):
Approving in advance, good job!

NOTE: need to address: https://github.com/OpenZeppelin/rust-contracts-stylus/pull/822#issuecomment-3338533773

sumitvekariya commented:
> LGTM!

Seems you have already updated Changelog.md Thanks for doing that, are we good to go?

bidzyyys reviewed (commented):
@sumitvekariya left comments how to improve your code as I cannot push to your repository.

sumitvekariya commented:
## Addressed Review Feedback

Thank you @bidzyyys for the thorough review! I've addressed all the feedback points:

### Changes Made

**1. Simplified `_safe_mint` calls**
- Removed redundant `.map_err(MethodError::encode)` in `deposit_for()` and `on_erc721_received()`
- Changed to simpler `?` operator

**2. Refactored error handling in `withdraw_to`**
- Updated to use `map_err` pattern instead of match statement with early returns
- Cleaner inline error mapping:
```rust
underlying
    .safe_transfer_from(...)
    .map_err(|e| match e {
        call::Error::Revert(reason) => reason,
        call::Error::AbiDecodingFailed(_) => vec![],
    })?;
```

**3. Replaced `.encode()` with `.abi_encode()` in tests**
- Updated all test assertions to use `.abi_encode()` directly on error structs
- Simplified error construction by removing wrapper `Error` enum:
  - `ERC721UnsupportedToken { ... }.abi_encode()`
  - `erc721::ERC721InvalidSender { ... }.abi_encode()`
  - `erc721::ERC721NonexistentToken { ... }.abi_encode()`
  - `erc721::ERC721InsufficientApproval { ... }.abi_encode()`

**4. Updated imports**
- Added `use alloy_sol_types::SolError;` for `abi_encode` method

### Test Results
✅ All 15 wrapper tests passing (2 ignored due to motsu limitations)  
✅ All 133 ERC721 tests passing  
✅ No linter errors

The code is now cleaner and more idiomatic. Ready for re-review! 🚀
